support/util/util.go

```diff
@@ -458,3 +458,109 @@ func PredicatesForHostedClusterChildResourcesAnnotationScoping(r client.Reader)
 	}
 	return predicate.NewPredicateFuncs(filter)
 }
+
+// PredicatesForNodepoolAnnotationScoping returns predicate filters for all event types that will ignore incoming
+// event requests for nodepool resources in which their owning hostedcluster resource doesn't have a scope annotation
+// that matches what is specified in the HOSTEDCLUSTERS_SCOPE_ANNOTATION env var.  If not defined or empty, the default behavior
+// is to accept all nodepool events in which the owning hostedcluster resource does not have a corresponding scope annotation defined.
+// The ENABLE_HOSTEDCLUSTERS_ANNOTATION_SCOPING env var must also be set to "true" to enable the scoping feature.
+func PredicatesForNodepoolAnnotationScoping(r client.Reader) predicate.Predicate {
+	hcAnnotationScopingEnabledEnvVal := os.Getenv(EnableHostedClustersAnnotationScopingEnv)
+	hcScopeAnnotationEnvVal := os.Getenv(HostedClustersScopeAnnotationEnv)
+	filter := func(obj client.Object) bool {
+		if hcAnnotationScopingEnabledEnvVal != "true" {
+			return true
+		}
+
+		np, ok := obj.(*hyperv1.NodePool)
+		if !ok {
+			return true
+		}
+
+		// use the Cluster Name from the nodepool spec to get the owning hostedcluster object
+		hostedClusterName := ""
+		if np.Spec.ClusterName != "" {
+			hostedClusterName = np.Spec.ClusterName
+		}
+		if hostedClusterName == "" {
+			return true
+		}
+		namespacedName := ParseNamespacedName(fmt.Sprintf("%s/%s", np.Namespace, hostedClusterName))
+		hcluster := &hyperv1.HostedCluster{}
+		err := r.Get(context.Background(), namespacedName, hcluster)
+		if err != nil {
+			return true
+		}
+
+		hostedClusterScopeAnnotation := ""
+		if hcluster.GetAnnotations() != nil {
+			hostedClusterScopeAnnotation = hcluster.GetAnnotations()[HostedClustersScopeAnnotation]
+		}
+		if hostedClusterScopeAnnotation == "" && hcScopeAnnotationEnvVal == "" {
+			return true
+		}
+		if hostedClusterScopeAnnotation != hcScopeAnnotationEnvVal {
+			return false // ignore event; the associated hostedcluster's scope annotation does not match what is defined in HOSTEDCLUSTERS_SCOPE_ANNOTATION
+		}
+		return true
+	}
+	return predicate.NewPredicateFuncs(filter)
+}
+
+// PredicatesForNodepoolChildResourcesAnnotationScoping returns predicate filters for all event types that will ignore incoming
+// event requests for resources in which the parent hostedcluster does not
+// match the "scope" annotation specified in the HOSTEDCLUSTERS_SCOPE_ANNOTATION env var.  If not defined or empty, the
+// default behavior is to accept all events for hostedclusters that do not have the annotation.
+// The ENABLE_HOSTEDCLUSTERS_ANNOTATION_SCOPING env var must also be set to "true" to enable the scoping feature.
+func PredicatesForNodepoolChildResourcesAnnotationScoping(r client.Reader) predicate.Predicate {
```

Comment:
Could you DRY out these helpers by passing in the function to extract the `HostedCluster`'s `NamespacedName` from the `obj client.Object`?

Reply:
What function are you referring to exactly?  Is there one that exists already somewhere or are you saying to create one use that?

Reply:
The latter - it seemed like the only difference between these utilities

Reply:
Nice! This looks really good!